File: Projeto-MobiFix/Backend/Models/Financas/FaturaDto.cs
namespace Backend.Models;

using System.ComponentModel.DataAnnotations;

public class FaturaDto
{
    public string NumeroFatura { get; set; } = string.Empty;
    public string ClienteNIF { get; set; } = string.Empty;
    public int? ServicoID { get; set; }
    public int? VendaID { get; set; }
    public decimal ValorTotal { get; set; }
    public string MetodoPagamento { get; set; } = string.Empty;
    public string? DataEmissao { get; set; }
    public List<DevolucaoDto> Devolucoes { get; set; } = new();
}




public class DevolucaoDto
{
    public string? DevolucaoID { get; set; }
    public string? DataDevolucao { get; set; }
    public string Motivo { get; set; } = string.Empty;
    public NotaCreditoDto? NotaCredito { get; set; }
}

public class DevolucaoCriacaoDto
{
    [Required]
    [MinLength(3, ErrorMessage = "O motivo deve ter pelo menos 3 caracteres.")]
    public string Motivo { get; set; } = string.Empty;
}

public class NotaCreditoDto
{
    public decimal ValorCreditado { get; set; }
}

public class FaturaCriacaoDto
{
    [Required]
    public string NumeroFatura { get; set; } = string.Empty;
    [Required]
    public string ClienteNIF { get; set; } = string.Empty;
    public int? ServicoID { get; set; }
    public int? VendaID { get; set; }
    
    [Range(0.01, double.MaxValue)]
    public decimal ValorTotal { get; set; }
    [Required]
    public string MetodoPagamento { get; set; } = string.Empty;

    // NOVO: Receber os itens para podermos abater no stock
    public List<ItemVendaCriacaoDto> ItensVenda { get; set; } = new();
}

File: Projeto-MobiFix/Database/dtos/financeDtos/faturaDto.js
const paraFaturaDto = (f) => {
    if (!f) return null;
    return {
        NumeroFatura: f._id.toString(),
        ClienteNIF: f.clienteId,
        ServicoID: f.servicoId || null,
        VendaID: f.vendaId || null,
        ValorTotal: f.valorTotal,
        MetodoPagamento: f.metodoPagamento,
        DataEmissao: f.dataEmissao,
        Devolucoes: (f.devolucoes || []).map(d => ({
            DevolucaoID: d._id?.toString() || null,
            DataDevolucao: d.dataDevolucao,
            Motivo: d.motivo,
            NotaCredito: d.notaCredito ? {
                ValorCreditado: d.notaCredito.valorCreditado
            } : null
        }))
    };
};

module.exports = { paraFaturaDto };
